packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+Contributions to the `@wordpress/components` package should follow a set of core principles and technical requirements.
 
-These examples can be consumed automatically from other projects in order to visualize them in their documentation. To ensure these examples are extractable, compilable and renderable, they should be structured in the following way:
+This set of guidelines should apply especially to newly introduced components. It is, in fact, possible that some of the older component don't respect some of these guidelines for legacy/compat reasons.
 
--   It has to be included in a `jsx` code block.
--   It has to work out-of-the-box. No additional code should be needed to have working the example.
--   It has to define a React component called `My<ComponentName>` which renders the example (i.e.: `MyButton`). Examples for the Higher Order Components should define a `MyComponent<ComponentName>` component (i.e.: `MyComponentWithNotices`).
+### Compatibility
+
+The `@wordpress/components` package includes components that are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes.
+
+In these situations, one possible approach is to "soft-deprecate" a given legacy API. This is achieved by:
+
+1. Removing traces of the API from the docs, while still supporing it in code;
+2. Updating all places in Gutenberg that use that API;
+3. Adding deprecation warnings (only after the previous point is completed, otherwide the Browser Console will be polluted by all those warnings and some e2e tests may fail).
+
+### Components composition
+
+<!-- ### Polymorphic Components (i.e. the `as` prop)
+
+The primary way to compose components is through the `as` prop. This prop can be used to change the underlying element used to render a component, e.g.:
+
+```tsx
+function LinkButton( { href, children } ) {
+	return <Button variant="primary" as="a" href={href}>{ children }</Button>;
+}
+```
+
+### Composition patterns
+
+TBD — E.g. Using `children` vs custom render props vs arbitrary "data" props
+
+### (Semi-)Controlled components
+
+TBD
+
+### Layout "responsibilities"
+
+TBD — Components' layout responsibilities and boundaries (i.e., a component should only affect the layout of its children, not its own) -->
+
+#### Components & Hooks
+
+One way to enable reusability and composition is to extract a component's underlying logic into a hook (living in a separate `hook.ts` file). The actual component (usually defined in a `component.tsx` file) can then invoke the hook and use its output to render the required DOM elements. For example:
+
+```tsx
+// in `hook.ts`
+function useExampleComponent( props: PolymorphicComponentProps< ExampleProps, 'div' > ) {
+	// Merge received props with the context system.
+	const { isVisible, className, ...otherProps	} = useContextSystem( props, 'Example' );
```

Comment:
```suggestion
	const { isVisible, className, ...otherProps } = useContextSystem( props, 'Example' );
```

Reply:
Is Prettier ignoring `tsx` in MD files?

Reply:
I don't personally remember Prettier ever auto-formatting code snippets in MD files — has that always been supposed to happen?

Reply:
Prettier is definitely able to do it: https://prettier.io/blog/2017/11/07/1.8.0.html

> Powered by Prettier's generic "multiparser", Prettier will format code blocks in Markdown! We use the language code provided with the code block to determine which language it is, and thus we can format any language that Prettier supports (including Markdown itself, if you're into that).

It might be slipping up on the `tsx` definition? Or could just be a configuration issue.

Reply:
It doesn't seem to work for me even for `jsx` snippets — does it work for you?